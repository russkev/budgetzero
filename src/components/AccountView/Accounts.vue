<template>
  <v-card elevation="0" class="mx-auto">
    <v-data-table
      id="accountsTable"
      :headers="headers"
      :items="accounts"
      sort-by="calories"
      class="elevation-1 account-table"
      hide-default-footer
      disable-pagination
    >
      <template #top>
        <v-toolbar flat color="white">
          <span class="text-h3">Accounts</span>
          <v-spacer />

          <AccountAddModal v-model="showModal" :edited-item="editedItem" @save="save()" />

          <v-btn id="addAccountBtn" color="accent" :dark="budgetExists" :disabled="!budgetExists" class="mb-2" @click="create()" >
            Add Account
          </v-btn>
        </v-toolbar>
        <v-divider class="pb-4" />
      </template>

      <template #item.action="{ item }">
        <div class="crud-actions">
          <v-icon icon dark class="" color="primary" @click="editItem(item)"> edit </v-icon>
          <v-icon icon dark class="ml-1" color="accent" @click="deleteItem(item)"> delete </v-icon>
        </div>
      </template>
      <template #no-data>
        <!-- <v-btn color="primary" @click="on">Add Account</v-btn> -->
      </template>
    </v-data-table>
  </v-card>
</template>



<script>
import { mapGetters } from 'vuex'
import AccountAddModal from './AccountAddModal.vue'
import { ID_NAME } from '../../constants'
// import { generateShortId } from "@/store/modules/id-module"

const DEFAULT_ACCOUNT_ITEM = {
  type: '',
  checkNumber: true,
  closed: false,
  name: '',
  note: null,
  sort: 0,
  onBudget: true,
  // balanceIsNegative: false,
  sign: 1,
  initialBalance: 0,
}
/*
Account view
-------------

+ Add account
+ Delete account
+ Edit account
    ++ Rename
    ++ TODO: Add plaid association

View: Name -- Type -- Balance

{
  "type": "CREDIT",         //type one of: MORTGAGE ASSET CREDIT DEBIT INVESTMENT SAVINGS CASH
  "checkNumber": true,      //checkNumber true if check column is enabled
  "closed": false,
  "name": "Mortgage",
  "note": null,
  "sort": 7,                //sort can just all be zero initially
  "onBudget": false,
  "_id": â€œb_{budgetId}_account_{accountId}
}

*/



export default {
  name: 'AccountGrid',
  components: {
    AccountAddModal
  },
  data() {
    return {
      headers: [
        {
          text: 'Name',
          align: 'left',
          sortable: false,
          value: 'name'
        },
        { text: 'Type', value: 'type' },
        { text: 'On Budget', value: 'onBudget' },
        { text: 'Invert Balance', value: 'sign' },
        { text: 'Closed', value: 'closed' },
        { text: 'Actions', value: 'action', sortable: false }
      ],
      editedIndex: -1,
      editedItem: null,
      previousItem: null,

      showModal: false,
      nameRules: [
        (v) => !!v || 'Name is required',
        (v) => (v && v.length <= 10) || 'Name must be less than 10 characters'
      ]
    }
  },
  computed: {
    ...mapGetters(['accounts', 'selectedBudgetId']),
    budgetExists() {
      return this.selectedBudgetId != "null"
    }
  },
  mounted() {},
  created() {},
  methods: {
    create() {
      this.editedIndex = -1
      this.editedItem = {...DEFAULT_ACCOUNT_ITEM}
      this.previousItem = null
      this.showModal = true
    },
    editItem(item) {
      this.editedIndex = this.accounts.indexOf(item)
      this.editedItem = Object.assign({}, item)
      this.previousItem = item
      this.showModal = true
    },
    deleteItem(item) {
      const index = this.accounts.indexOf(item)
      this.$root
        .$confirm('Delete Account', 'Are you sure you want to delete this account?', {
          onlyShowAgreeBtn: false,
          agreeBtnColor: 'accent',
          agreeBtnText: 'Delete Account',
          cancelBtnColor: 'grey'
        })
        .then((confirm) => {
          this.$store
            .dispatch('deleteAccount', this.accounts[index])
            .then((response) => {
              console.log('Got some data, now lets show something in this component', response)
            })
            .catch((error) => {
              // Action failed
              this.$root.$confirm(
                'Deletion Failed',
                `This account still has ${error} transaction(s). You must delete those transactions to delete the account.`,
                {
                  onlyShowAgreeBtn: true,
                  agreeBtnColor: 'accent',
                  agreeBtnText: 'Ok'
                }
              )
            })
        })
    },
    close() {
      this.showModal = false
      setTimeout(() => {
        this.resetItems()
      }, 300)
    },
    save() {
      if (this.editedIndex > -1) {
        const payload = {
          ...this.editedItem
        }
        this.$store
          .dispatch('commitDocToPouchAndVuex', { current: payload, previous: this.previousItem})
          .then(() => {
            this.$store.dispatch('calculateAllValues')
          })
      } else {
        const payload = {
          ...this.editedItem,
          _id: `b_${this.selectedBudgetId}${ID_NAME.account}${this.generateShortId()}`
        }
        this.$store
          .dispatch('commitDocToPouchAndVuex', {current: payload, previous: null})
      }
      
    },
    resetItems() {
      this.editedIndex = -1
      this.editedItem = {...DEFAULT_ACCOUNT_ITEM}
      this.previousItem = null
    },

    
  }
}
</script>

<style scoped></style>
