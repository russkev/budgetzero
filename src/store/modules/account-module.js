import { sanitizeValueInput } from '../../helper'
import moment from 'moment'
import { DEFAULT_ACCOUNT_BALANCE, ID_LENGTH, ID_NAME } from '../../constants'
import _ from 'lodash'
import Vue from 'vue'

const DEFAULT_ACCOUNT_STATE = {
  allAccountBalances: {},
  accounts: [],
  intlCurrency: new Intl.NumberFormat("en-us", { style: "currency", currency: "USD" }),
}

export default {
  state: {
    ...DEFAULT_ACCOUNT_STATE
  },
  getters: {
    allAccountBalances: (state) => state.allAccountBalances,
    accounts: (state) => state.accounts,
    accountsById: (state) => {
      return state.accounts.reduce((partial, account) => {
        partial[account._id.slice(-ID_LENGTH.account)] = account
        return partial
      }, {})
    },
    accountsOnBudget: (state) => {
      return state.accounts.filter((account) => account.onBudget)
    },
    accountsOffBudget: (state) => {
      return state.accounts.filter((account) => !account.onBudget)
    },
    intlCurrency: (state) => state.intlCurrency,
  },
  mutations: {
    SET_ACCOUNTS(state, accounts) {
      state.accounts = accounts
      accounts.forEach((account) => {
        const id = account._id.slice(-ID_LENGTH.account)
        if (state.allAccountBalances[id] === undefined) {
          Vue.set(state.allAccountBalances, id, DEFAULT_ACCOUNT_BALANCE)
        }
      })
    },
    SET_ALL_ACCOUNT_BALANCES(state, accountBalances) {
      Object.entries(accountBalances).forEach(([account_id, balances]) => {
        Vue.set(state.allAccountBalances, account_id, balances)
      })
    },
    UPDATE_ACCOUNT_BALANCES(state, { account, account_id, cleared, uncleared, working }) {
      if (state.allAccountBalances[account_id] === undefined) {
        Vue.set(state.allAccountBalances, account_id, DEFAULT_ACCOUNT_BALANCE)
      }
      updateAccountBalances(state.allAccountBalances, account, account_id, cleared, uncleared, working)
    },
    RESET_ACCOUNT_STATE(state) {
      Object.entries(DEFAULT_ACCOUNT_STATE).forEach(([key, value]) => {
        Vue.set(state, key, value)
      })
    }
  },
  actions: {
    createUpdateAccount(context, payload) {
      context.dispatch('commitDocToPouchAndVuex', { current: payload.account, previous: null }).then((response) => {
        const date = moment(new Date()).format('YYYY-MM-DD')
        if (payload.initialBalance) {
          const initTransaction = {
            account: response._id.slice(-ID_LENGTH.account),
            category: null,
            cleared: true,
            approved: true,
            value: sanitizeValueInput(payload.initialBalance) * 100,
            date: date,
            memo: null,
            reconciled: true,
            flag: '#ffffff',
            payee: `---------------------initial-balance`,
            transfer: null,
            splits: [],
            _id: `b_${context.getters.selectedBudgetId}${ID_NAME.transaction}${this._vm.generateId(date)}`
          }
          return context.dispatch('createOrUpdateTransaction', { current: initTransaction, previous: null })
        }

        return
      })
    },
    deleteAccount(context, payload) {
      return new Promise((resolve, reject) => {
        const myId = payload._id.slice(-ID_LENGTH.account)
        this._vm.$pouch
          .query((doc, emit) => {
            if (doc.account === myId) {
              emit(doc)
            }
          })
          .then((result2) => {
            if (result2.total_rows > 0) {
              // Account still has transactions, so resolve with amount of transactions in account for error message.
              reject(result2.total_rows)
            } else {
              // Dispatch account for deletion
              context.dispatch('commitDocToPouchAndVuex', { current: null, previous: payload })
              resolve('Success')
            }
          })
          .catch((err) => {
            reject('Error')
            console.log(err)
          })
      })
    }
  }
}

const updateAccountBalances = (current_balances, account, account_id, cleared, uncleared, working) => {
  let updated_balances = {...current_balances[account_id]}
  updated_balances.cleared += cleared * account.sign
  updated_balances.uncleared += uncleared * account.sign
  updated_balances.working += working * account.sign
  if (working * account.sign > 0) {
    updated_balances.income += working * account.sign
  } else {
    updated_balances.expense += working * account.sign * -1
  }

  Vue.set(current_balances, account_id, updated_balances)
}

export { updateAccountBalances }
