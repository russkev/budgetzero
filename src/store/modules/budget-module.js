import _ from 'lodash'
import moment from 'moment'
import Vue from 'vue'

import { ID_LENGTH, ID_NAME } from '../../constants.js'

const DEFAULT_BUDGET_STATE = {
  allBudgets: [],
  budgetBalances: {},
  budgetExists: false, // This opens the create budget modal when 'false'
  isLoadingFullscreen: true,
  targetPage: null
}

export default {
  state: {
    ...DEFAULT_BUDGET_STATE
  },
  getters: {
    allBudgets: (state) => state.allBudgets,
    budgetsById: (state) => {
      return state.allBudgets.reduce((partial, budget) => {
        if (budget._id !== undefined) {
          partial[budget._id.slice(-ID_LENGTH.budget)] = budget
        }
        return partial
      }, {})
    },
    budgetExists: (state) => state.budgetExists,
    budgetBalances: (state) => state.budgetBalances,
    isLoadingFullscreen: (state) => state.isLoadingFullscreen,
    targetPage: (state) => state.targetPage
  },
  mutations: {
    RESET_BUDGET_STATE(state) {
      Object.entries(DEFAULT_BUDGET_STATE).forEach(([key, value]) => {
        Vue.set(state, key, value)
      })
    },
    SET_ALL_BUDGETS(state, budgets) {
      if (budgets.length == 0) {
        Vue.set(state, 'budgetExists', false)
      } else {
        Vue.set(state, 'budgetExists', true)
      }
      Vue.set(state, 'allBudgets', budgets)
    },
    SET_LOADING_FULLSCREEN(state, isLoading) {
      Vue.set(state, 'isLoadingFullscreen', isLoading)
    },
    SET_TARGET_PAGE(state, targetPage) {
      Vue.set(state, 'targetPage', targetPage)
    },
    RESET_BUDGETS(state) {
      Vue.set(state, 'allBudgets', [])
      Vue.set(state, 'budgetExists', false)
    }
  },
  actions: {
    /**
     * Creates new budget and commits to pouchdb
     * @param {*} context
     * @param {string} budgetName The name of the budget to be created
     */
    createBudget: async ({ commit, dispatch }, { name, masterCategories, categories }) => {
      if (!name) {
        commit('SET_SNACKBAR_MESSAGE', {
          snackbarMessage: 'Invalid budget name',
          snackbarColor: 'error'
        })
        Promise.reject(`Invalid budget name: ${name}`)
      }
      const budget_id = await dispatch('generateUniqueShortId', { prefix: ID_NAME.budget })

      const budget = {
        name: name,
        currency: 'USD',
        created: moment(new Date()).format('YYYY-MM-DD'),
        accessed: Math.floor(Date.now() / 1000),
        checkNumber: false,
        _id: `${ID_NAME.budget}${budget_id}`
      }

      return dispatch('createLocalPouchDB')
        .then(() => {
          return dispatch('commitDocToPouchAndVuex', { current: budget, previous: null })
        })
        .then(() => {
          return dispatch('loadLocalBudget')
        })
        .then(() => {
          return dispatch('initializeBudgetCategories', { masterCategories, categories })
        })
        .then(() => {
          dispatch('getAllDocsFromPouchDB')
        })
        .catch((err) => {
          console.log(err)
        })
    },

    commitBudgetToVuex(context, { current, previous }) {
      if ((current && !current._id) || (previous && !previous._id)) {
        console.error(`commitBudgetToVuex called with invalid arguments, current: ${current}, previous: ${previous}`)
        return
      }
      const new_budget_id = current ? current._id.slice(-ID_LENGTH.budget) : previous._id.slice(-ID_LENGTH.budget)

      return context.dispatch('setSelectedBudgetID', new_budget_id)
    },

    updateSelectedBudgetId(context, budgets) {
      let selected_budget_id = context.getters.selectedBudgetId

      if (selected_budget_id === null || context.getters.budgetsById[selected_budget_id] === undefined) {
        const local_storage_budget_id = localStorage.getItem('budgetID')
        if (local_storage_budget_id && budgets.map((budget) => budget._id).indexOf(local_storage_budget_id) > -1) {
          selected_budget_id = local_storage_budget_id
        } else if (budgets.length > 0) {
          selected_budget_id = budgets[0]._id.slice(-ID_LENGTH.budget)
        }
      }

      if (selected_budget_id !== null) {
        context.commit('UPDATE_SELECTED_BUDGET_ID', selected_budget_id)
      }
      return selected_budget_id
    },

    async updateBudgetAccessed({ getters, dispatch }, budget_id) {
      const db = this._vm.$pouch
      let budget = getters.budgetsById[budget_id]
      if (budget === undefined) {
        return
      }
      return db
        .put({
          ...budget,
          accessed: Math.floor(Date.now() / 1000)
        })
        .then(() => {
          return dispatch('fetchAllBudgets')
        })
    },

    /**
     * Deletes entire budget
     * @param {*} context
     * @param {*} payload budget_ document
     */
    deleteEntireBudget(context, budget_document) {
      const budget_id = budget_document._id.slice(-ID_LENGTH.budget)
      const budget_name = budget_document.name

      if (!budget_id || !budget_document) {
        return
      }

      return Promise.all([
        context.dispatch('deleteDocFromPouch', budget_document),
        context.dispatch('deleteAllBudgetDocuments', budget_id)
      ])
        .then((results) => {
          if (results[0].ok) {
            context.commit('SET_SNACKBAR_MESSAGE', `${budget_name} has been deleted`)
            return true
          }
          return false
        })
        .catch((error) => {
          context.commit('API_FAILURE', error)
        })
    },

    resetBudget(context) {
      context.commit('RESET_BUDGET_STATE')
    },

    /**
     * Turn the right-red-arrow on/off
     */
    flipOverspending(context, item) {
      var payload = {
        budget: 0,
        overspending: true,
        note: '',
        _id: `b_${context.getters.selectedBudgetId}${ID_NAME.monthCategory}${
          context.getters.month_selected
        }_${item._id.slice(-ID_LENGTH.category)}`,
        date: context.getters.month_selected + '-01'
      }

      //Check if already exists
      if (
        context.getters.monthCategoryBudgets[context.getters.month_selected] &&
        context.getters.monthCategoryBudgets[context.getters.month_selected][item._id.slice(-ID_LENGTH.category)]
      ) {
        payload = JSON.parse(
          JSON.stringify(
            context.getters.monthCategoryBudgets[context.getters.month_selected][item._id.slice(-ID_LENGTH.category)]
          )
        )

        payload.overspending = !payload.overspending
      }
      context.dispatch('updateMonthCategory', payload)
    }
  }
}
