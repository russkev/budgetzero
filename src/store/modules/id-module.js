import { ntob } from 'number-to-base64'
import Vue from 'vue'
import { randomInt } from '../../helper'


export default {
  state: {},
  getters: {},
  mutations: {},
  actions: {
    /**
     * Generate a short ID and unsure it doesn't already exist in the database
     * @param {*} context
     * @returns
     */
    generateUniqueShortId: async (context, { prefix }) => {
      let unique_id = false
      let num_tries = 0
      const max_tries = 10
      while (!unique_id && num_tries < max_tries) {
        num_tries += 1
        const id = this.generateShortId()
        const id_exists = await idExists(prefix + id)
        if (!id_exists) {
          unique_id = id
        }
      }
      if (unique_id) {
        return unique_id
      } else {
        context.commit('SET_SNACKBAR_MESSAGE', {
          snackbarMessage: 'Unable to create unique ID',
          snackbarColor: 'error'
        })
        Promise.reject(`Unable to create unique ID, tried ${max_tries} times`)
      }
    }
  }
}

/**
 * Return true if id already exists in database, false otherwise
 * @param {*} context
 * @param {String} id
 * @returns
 */
function idExists(id) {
  const db = Vue.prototype.$pouch
  return db
    .get(id)
    .then(() => {
      return true
    })
    .catch(() => {
      return false
    })
}

function validateId(id) {
  return id.length == 3 || id.length == 12
}

function daySeconds() {
  const date = new Date()
  return date.getSeconds() + 60 * date.getMinutes() + 60 * 60 * date.getHours()
}

function daySecondsBase64() {
  let day_seconds_64 = urlSafeBase64(daySeconds())
  // while (day_seconds_64.length < 4) {
  //   // day_seconds_64 += urlSafeCharacter()
  //   day_seconds_64 
  // }
  return day_seconds_64.padStart(4, '-')
}

function isValidDate(date_string) {
  if (!(typeof date_string === 'string' || date_string instanceof String)) {
    return false
  }
  return /^[0-9]{4}\-[0-9]{2}\-[0-9]{2}$/.test(date_string)
}

const original_encoding = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/='
const new_encoding      = '-.0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz'
const encode_mapping = original_encoding.split('').reduce((partial, letter, i) => {
  partial[letter] = new_encoding[i]
  return partial
}, {})

function urlSafeBase64(decimal_number, length = null) {
  let result = ntob(decimal_number)
  // console.log(`NTOB RESULT: ${result}`)
  // console.log(encode_mapping)
  result = result
    .split('')
    .map((character) => {
      return encode_mapping[character]
    })
    .join('')

  if (length) {
    while (result.length < length) {
      result += urlSafeCharacter()
    }
    return result.slice(0, length)
  }
  return result
}

function urlSafeCharacter() {
  const characters = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-_.'
  return characters[randomInt(0, characters.length - 1)]
}

function urlSafeString(length) {
  let result = ''
  while (result.length < length) {
    result += urlSafeCharacter()
  }
  return result
}

function base64Date(date) {
  if (!date || !isValidDate(date)) {
    date = new Date().toISOString().split('T')[0]
  }
  const date_number = parseInt(date.split('-').join(''), 10)
  return urlSafeBase64(date_number).padStart(4, '-')
}
/**
 * Creates a ID of 12 characters encoded in base 64:
 *   1 -  4: Year month day
 *   5 -  8: Number of seconds so far in the day
 *   9 - 12: Random string / transaction ID
 *   Note: A base 64 character string of length 4 has over
 *   16 million combinations.
 *
 * @param {string} date In YYYY-MM-DD
 * @returns 12 Character string
 */
function generateId(date = null, transaction_id = null) {
  const date_encoded = base64Date(date)
  const seconds_encoded = daySecondsBase64()

  let id_encoded = ''
  try {
    id_int = parseInt(transaction_id)
    id_encoded = urlSafeBase64(id_int)
  } catch {
    id_encoded = urlSafeString(4)
  }

  return date_encoded + seconds_encoded + id_encoded
}

function generateShortId() {
  return urlSafeString(3)
}

export { generateId, generateShortId, validateId, base64Date }
