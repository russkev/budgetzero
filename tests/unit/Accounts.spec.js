import { mount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import Accounts from "@/components/AccountView/Accounts.vue";
import Vuetify from "vuetify";
import mock_budget from "@/../tests/__mockdata__/mock_budget_2.json";
import Vue from "vue";
import { generateShortId } from '@/store/modules/id-module'

Vue.use(Vuetify);

const localVue = createLocalVue();
localVue.use(Vuex);

const $route = {
  path: '/fake/path',
  params: {
    account_id: '6JK'
  }
}

const data = mock_budget.rows.map(row => row.doc);

describe("accounts table", () => {
  let vuetify;
  let wrapper;
  let getters;
  let store;

  beforeEach(() => {
    vuetify = new Vuetify();
    localVue.use(vuetify);
    localVue.use(Vuex)

    getters = {
      selectedBudgetId: () => 'rkW',
      accounts: () => data.filter(row => row._id.includes("_account_")),
    }

    store = new Vuex.Store({
      getters
    })

    wrapper = mount(Accounts, {
      store,
      localVue,
      vuetify,
      generateShortId,
      mocks: {
        $route,
      },
      stubs: {
        ImportModalComponent: true
      }
    })
  });

  it("account renders correct snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("account table exists", () => {
    expect(wrapper.find("#accountsTable").exists()).toBeTruthy();
  });

  it("clicking Add Account button shows modal", async () => {
    // Open modal
    await wrapper.find("#addAccountBtn").trigger("click");

    // Verify modal is open
    expect(wrapper.find(".v-dialog .title").text()).toEqual("Add Account");
  });

  it("create new account", async () => {
    // Open modal
    await wrapper.find("#addAccountBtn").trigger("click");

    // //Mocks & setup
    wrapper.vm.$store.dispatch = vi.fn();
    wrapper.vm.generateShortId = vi.fn()
    wrapper.vm.generateShortId.mockReturnValue('a3-')

    await wrapper.find("#nameField").setValue("nameofnewAccount");
    await wrapper.find("#typeField").setValue("CHECKING");
    await wrapper.find("#noteField").setValue("test note");

    await wrapper.find("#saveAccountBtn").trigger("click");

    expect(wrapper.vm.$store.dispatch).toBeCalled()
    expect(wrapper.vm.$store.dispatch).toBeCalledWith('commitDocToPouchAndVuex', {
      current: {
        _id: 'b_rkW_account_a3-',
        type: 'CHECKING',
        checkNumber: true,
        closed: false,
        name: 'nameofnewAccount',
        note: 'test note',
        sort: 0,
        onBudget: true,
        sign: 1,
        initialBalance: 0
      },
      previous: null
    })
  });
});
