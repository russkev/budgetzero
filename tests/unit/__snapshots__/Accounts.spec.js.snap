// Vitest Snapshot v1

exports[`accounts table > account renders correct snapshot 1`] = `
"<div class=\\"mx-auto v-card v-sheet theme--light elevation-0\\">
  <div class=\\"v-data-table elevation-1 account-table v-data-table--has-top theme--light\\" id=\\"accountsTable\\">
    <header class=\\"v-sheet theme--light v-toolbar v-toolbar--flat white\\" style=\\"height: 64px;\\">
      <div class=\\"v-toolbar__content\\" style=\\"height: 64px;\\"><span class=\\"text-h3\\">Accounts</span>
        <div class=\\"spacer\\"></div>
        <div class=\\"v-dialog__container\\">
          <!---->
        </div><button type=\\"button\\" class=\\"mb-2 v-btn v-btn--is-elevated v-btn--has-bg theme--dark v-size--default accent\\" id=\\"addAccountBtn\\"><span class=\\"v-btn__content\\"> Add Account </span></button></div>
    </header>
    <hr role=\\"separator\\" aria-orientation=\\"horizontal\\" class=\\"pb-4 v-divider theme--light\\">
    <div class=\\"v-data-table__wrapper\\">
      <table>
        <colgroup>
          <col class=\\"\\">
          <col class=\\"\\">
          <col class=\\"\\">
          <col class=\\"\\">
          <col class=\\"\\">
          <col class=\\"\\">
        </colgroup>
        <thead class=\\"v-data-table-header\\">
          <tr>
            <th role=\\"columnheader\\" scope=\\"col\\" aria-label=\\"Name\\" class=\\"text-left\\"><span>Name</span></th>
            <th role=\\"columnheader\\" scope=\\"col\\" aria-label=\\"Type: Not sorted. Activate to sort ascending.\\" aria-sort=\\"none\\" class=\\"text-start sortable\\"><span>Type</span><i aria-hidden=\\"true\\" class=\\"v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light\\" style=\\"font-size: 18px;\\"></i></th>
            <th role=\\"columnheader\\" scope=\\"col\\" aria-label=\\"On Budget: Not sorted. Activate to sort ascending.\\" aria-sort=\\"none\\" class=\\"text-start sortable\\"><span>On Budget</span><i aria-hidden=\\"true\\" class=\\"v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light\\" style=\\"font-size: 18px;\\"></i></th>
            <th role=\\"columnheader\\" scope=\\"col\\" aria-label=\\"Invert Balance: Not sorted. Activate to sort ascending.\\" aria-sort=\\"none\\" class=\\"text-start sortable\\"><span>Invert Balance</span><i aria-hidden=\\"true\\" class=\\"v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light\\" style=\\"font-size: 18px;\\"></i></th>
            <th role=\\"columnheader\\" scope=\\"col\\" aria-label=\\"Closed: Not sorted. Activate to sort ascending.\\" aria-sort=\\"none\\" class=\\"text-start sortable\\"><span>Closed</span><i aria-hidden=\\"true\\" class=\\"v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light\\" style=\\"font-size: 18px;\\"></i></th>
            <th role=\\"columnheader\\" scope=\\"col\\" aria-label=\\"Actions\\" class=\\"text-start\\"><span>Actions</span></th>
          </tr>
        </thead>
        <tbody>
          <tr class=\\"\\">
            <td class=\\"text-left\\">Credit</td>
            <td class=\\"text-start\\">CREDIT</td>
            <td class=\\"text-start\\">true</td>
            <td class=\\"text-start\\">-1</td>
            <td class=\\"text-start\\">false</td>
            <td class=\\"text-start\\">
              <div class=\\"crud-actions\\"><button type=\\"button\\" icon=\\"\\" class=\\"v-icon notranslate v-icon--link material-icons theme--dark primary--text\\">edit</button><button type=\\"button\\" icon=\\"\\" class=\\"v-icon notranslate ml-1 v-icon--link material-icons theme--dark accent--text\\">delete</button></div>
            </td>
          </tr>
          <tr class=\\"\\">
            <td class=\\"text-left\\">Investments</td>
            <td class=\\"text-start\\">INVESTMENT</td>
            <td class=\\"text-start\\">true</td>
            <td class=\\"text-start\\">1</td>
            <td class=\\"text-start\\">false</td>
            <td class=\\"text-start\\">
              <div class=\\"crud-actions\\"><button type=\\"button\\" icon=\\"\\" class=\\"v-icon notranslate v-icon--link material-icons theme--dark primary--text\\">edit</button><button type=\\"button\\" icon=\\"\\" class=\\"v-icon notranslate ml-1 v-icon--link material-icons theme--dark accent--text\\">delete</button></div>
            </td>
          </tr>
          <tr class=\\"\\">
            <td class=\\"text-left\\">Savings</td>
            <td class=\\"text-start\\">SAVING</td>
            <td class=\\"text-start\\">true</td>
            <td class=\\"text-start\\">1</td>
            <td class=\\"text-start\\">false</td>
            <td class=\\"text-start\\">
              <div class=\\"crud-actions\\"><button type=\\"button\\" icon=\\"\\" class=\\"v-icon notranslate v-icon--link material-icons theme--dark primary--text\\">edit</button><button type=\\"button\\" icon=\\"\\" class=\\"v-icon notranslate ml-1 v-icon--link material-icons theme--dark accent--text\\">delete</button></div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>"
`;

exports[`accounts table account renders correct snapshot 1`] = `
<div class="mx-auto v-card v-sheet theme--light elevation-0">
  <div class="v-data-table elevation-1 account-table v-data-table--has-top theme--light" id="accountsTable">
    <header class="v-sheet theme--light v-toolbar v-toolbar--flat white" style="height: 64px;">
      <div class="v-toolbar__content" style="height: 64px;"><span class="text-h3">Accounts</span>
        <div class="spacer"></div>
        <div class="v-dialog__container">
          <!---->
        </div> <button type="button" class="mb-2 v-btn v-btn--is-elevated v-btn--has-bg theme--dark v-size--default accent" id="addAccountBtn"><span class="v-btn__content">
          Add Account
        </span></button>
      </div>
    </header>
    <hr role="separator" aria-orientation="horizontal" class="pb-4 v-divider theme--light">
    <div class="v-data-table__wrapper">
      <table>
        <colgroup>
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
        </colgroup>
        <thead class="v-data-table-header">
          <tr>
            <th role="columnheader" scope="col" aria-label="Name" class="text-left"><span>Name</span></th>
            <th role="columnheader" scope="col" aria-label="Type: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Type</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="On Budget: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>On Budget</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Invert Balance: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Invert Balance</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Closed: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Closed</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Actions" class="text-start"><span>Actions</span></th>
          </tr>
        </thead>
        <tbody>
          <tr class="v-data-table__empty-wrapper">
            <td colspan="6">No data available</td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
